source/extensions/common/async_files/async_file_context_thread_pool.cc

```diff
@@ -178,61 +188,69 @@ class ActionDuplicateFile : public AsyncFileActionThreadPool<absl::StatusOr<Asyn
     return std::make_shared<AsyncFileContextThreadPool>(context()->manager(), newfd.return_value_);
   }
 
-  void onCancelledBeforeCallback(absl::StatusOr<AsyncFileHandle> result) override {
-    if (result.ok()) {
-      result.value()->close([](absl::Status) {}).IgnoreError();
+  void onCancelledBeforeCallback() override {
+    if (result_.value().ok()) {
+      result_.value().value()->close(nullptr, [](absl::Status) {}).IgnoreError();
     }
   }
+  bool hasActionIfCancelledBeforeCallback() const override { return true; }
 };
 
 } // namespace
 
-absl::StatusOr<CancelFunction>
-AsyncFileContextThreadPool::stat(std::function<void(absl::StatusOr<struct stat>)> on_complete) {
-  return checkFileAndEnqueue(std::make_shared<ActionStat>(handle(), std::move(on_complete)));
+absl::StatusOr<CancelFunction> AsyncFileContextThreadPool::stat(
+    Event::Dispatcher* dispatcher,
```

Comment:
Yes - if it was captured in the file context then it would become impossible to pass a file handle between threads, which especially with the duplicate action might be desirable. The only real advantage to keeping the dispatcher in the class would be that it would have made this PR a lot smaller.

(Also, I regret naming this class `AsyncFileContextThreadPool` - at the time it seemed clear that it was an AsyncFileContext that performed its operations on the back of a thread pool, but now it keeps looking like the object itself is a thread pool which it very much is not.)